### PR TITLE
Add labels to pools

### PR DESCRIPTION
**Pull request by arthurcgc**
feat: add nodeAffinity according to pool label to pods when deploying in kubernetes
feat: use pool nodeSelector if cluster does not specify a node selector
change: don't use any nodeSelector if one is not found in cluster or pool

### CODE REVIEW

File: provision/kubernetes/deploy.go [hunk 1]
```diff
@@ -502,6 +503,41 @@ func ensureServiceAccountForApp(ctx context.Context, client *ClusterClient, a pr
...
+func podAffinity(ctx context.Context, a provision.App) (*apiv1.Affinity, error) {
+	pool, err := pool.GetPoolByName(ctx, a.GetPool())
+	if err != nil {
+		return nil, err
+	}
+
+	return pool.GetAffinity()
+}
+
 func createAppDeployment(ctx context.Context, client *ClusterClient, depName string, oldDeployment *appsv1.Deployment, a provision.App, process string, version appTypes.AppVersion, replicas int, labels *provision.LabelSet, selector map[string]string) (*appsv1.Deployment, *provision.LabelSet, error) {
```
Comment (arthurcgc) [inline provision/kubernetes/deploy.go:541]: add test that references pool.GetNodeSelector and pool.GetAffinnity

File: provision/pool/pool.go [hunk 2]
```diff
@@ -53,12 +61,40 @@ type AddPoolOptions struct {
...
 
 type UpdatePoolOptions struct {
 	Default *bool
 	Public  *bool
 	Force   bool
+
+	Labels map[string]string
+}
+
+func (p *Pool) GetNodeSelector() (map[string]string, error) {
```
Comment (arthurcgc) [inline provision/pool/pool.go:76]: test GetNodeSelector

File: provision/pool/pool.go [hunk 3]
```diff
@@ -53,12 +61,40 @@ type AddPoolOptions struct {
...
+		if err := json.Unmarshal([]byte(affinity), &k8sAffinity); err != nil {
+			return nil, err
+		}
+		return &k8sAffinity, nil
+	}
+
+	return nil, nil
 }
 
 func (p *Pool) GetProvisioner() (provision.Provisioner, error) {
```
Comment (arthurcgc) [inline provision/pool/pool.go:100]: add labels test case

File: provision/pool/pool.go [hunk 4]
```diff
@@ -305,6 +342,24 @@ func (p *Pool) MarshalJSON() ([]byte, error) {
 	return json.Marshal(&result)
 }
 
+func validateLabels(labels map[string]string) error {
```
Comment (arthurcgc) [inline provision/pool/pool.go:345]: create validate labels test

File: provision/pool/pool.go [hunk 5]
```diff
@@ -305,6 +342,24 @@ func (p *Pool) MarshalJSON() ([]byte, error) {
...
+		var affinity apiv1.Affinity
+		if err := json.Unmarshal([]byte(affinityStr), &affinity); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
 func (p *Pool) validate() error {
```
Comment (arthurcgc) [inline provision/pool/pool.go:363]: add labels to validate() test

File: provision/pool/pool.go [hunk 6]
```diff
@@ -315,11 +370,15 @@ func (p *Pool) validate() error {
...
 		return &tsuruErrors.ValidationError{Message: msg}
 	}
+
+	if len(p.Labels) > 0 {
+		validateLabels(p.Labels)
+	}
 	return nil
 }
 
 func AddPool(ctx context.Context, opts AddPoolOptions) error {
```
Comment (arthurcgc) [inline provision/pool/pool.go:380]: include labels to AddPoolTest

File: provision/pool/pool.go [hunk 7]
```diff
@@ -53,12 +61,40 @@ type AddPoolOptions struct {
...
+		err := json.Unmarshal([]byte(values), &nodeSelectorValues)
+		if err != nil {
+			return nil, err
+		}
+		return nodeSelectorValues, nil
+	}
+	return nil, nil
+}
+
+func (p *Pool) GetAffinity() (*apiv1.Affinity, error) {
```
Comment (arthurcgc) [inline provision/pool/pool.go:88]: test GetAffinity

File: provision/pool/pool.go [hunk 8]
```diff
@@ -563,6 +622,14 @@ func PoolUpdate(ctx context.Context, name string, opts UpdatePoolOptions) error
 	if opts.Default != nil {
 		query["default"] = *opts.Default
 	}
+	if len(opts.Labels) > 0 {
```
Comment (arthurcgc) [inline provision/pool/pool.go:625]: add pool update with labels test case

File: provision/kubernetes/cluster.go [hunk 9]
```diff
@@ -35,20 +35,21 @@ import (
...
+	namespaceLabelsKey            = "namespace-labels"
+	externalPolicyLocalKey        = "external-policy-local"
+	routerAddressLocalKey         = "router-local"
+	disableHeadlessKey            = "disable-headless"
+	maxSurgeKey                   = "max-surge"
+	maxUnavailableKey             = "max-unavailable"
+	singlePoolKey                 = "single-pool"
+	ephemeralStorageKey           = "ephemeral-storage"
+	preStopSleepKey               = "pre-stop-sleep"
+	disableDefaultNodeSelectorKey = "disable-default-node-selector"
```
Comment (wpjunior) [inline provision/kubernetes/cluster.go:52]: Great @arthurcgc

File: provision/kubernetes/cluster.go [hunk 10]
```diff
@@ -59,20 +60,21 @@ const (
...
+		disableHeadlessKey:            "Disable headless service creation for every app-process. This config may be prefixed with `<pool-name>:`.",
+		maxSurgeKey:                   "Max surge for deployments rollout. This config may be prefixed with `<pool-name>:`. Defaults to 100%.",
+		maxUnavailableKey:             "Max unavailable for deployments rollout. This config may be prefixed with `<pool-name>:`. Defaults to 0.",
+		singlePoolKey:                 "Set to use entire cluster to a pool instead only designated nodes. Defaults do false.",
+		ephemeralStorageKey:           fmt.Sprintf("Sets limit for ephemeral storage for created pods. This config may be prefixed with `<pool-name>:`. Defaults to %s.", defaultEphemeralStorageLimit.String()),
+		preStopSleepKey:               fmt.Sprintf("Number of seconds to sleep in the preStop lifecycle hook. This config may be prefixed with `<pool-name>:`. Defaults to %d.", defaultPreStopSleepSeconds),
+		disableDefaultNodeSelectorKey: "Disables the use of node selector in the cluster if set to true",
```
Comment (pedrokiefer) [inline provision/kubernetes/cluster.go:77]: `if enabled` instead of `set to true`

File: provision/pool/pool.go [hunk 11]
```diff
@@ -53,12 +60,28 @@ type AddPoolOptions struct {
...
 	Public  *bool
 	Force   bool
+
+	Labels map[string]string
+}
+
+func (p *Pool) GetAffinity() (*apiv1.Affinity, error) {
+	if affinity, ok := p.Labels[affinityKey]; ok {
+		var k8sAffinity apiv1.Affinity
+		if err := json.Unmarshal([]byte(affinity), &k8sAffinity); err != nil {
```
Comment (wpjunior) [inline provision/pool/pool.go:78]: @arthurcgc: nice to have: also support `yaml.Unmarshal` first and if an error tries json.
@pedrokiefer what do you think about ?

Comment (pedrokiefer) [reply provision/pool/pool.go:78]: I'm not sure. Maybe only yaml?The question is how we want to expose this in the terraform provider? Yaml is more kubernetes like.

Comment (arthurcgc) [reply provision/pool/pool.go:78]: I think only one mode (yaml or json) should suffice... If we're thinking of using it with terraform we should maintain the pattern we're implementing in the terraform-provider, which I'm not sure what it is xD

**Review by wpjunior — APPROVED** [review]
Thanks @arthurcgc

**Comment by wpjunior** [comment]
Thanks @arthurcgc, I am so glad to work at your side, keep hacking well <3

**Pull request merged**